### PR TITLE
feat(throttle): cap the requests for the heartbeat endpoint

### PR DESCRIPTION
**Pull request by Daniel-Boll**
## First Approach

This commit introduces a throttling mechanism to the heartbeat endpoint.
A new `ThrottleState` struct is added to the `AppState` to keep track of
the last request time for each user-channel pair. If a user tries to send
a heartbeat for the same channel within a minute, a 429 Too Many Requests
response is returned. The throttling state is shared across all requests
using an `Arc<Mutex<ThrottleState>>`.

Signed-off-by: Daniel Boll <danielboll.academico@gmail.com>

## Second Approach

After realizing that would be quite similar to use ScyllaDB for this task, we did a new table for Throttling using TTL in the insertion of the query. 

The partition key is made of:
- `request_uri`: /some/route
- `user_id`: user related to the Bearer Token sent in the request
- `content`: a random string that will be used to differentiate this request from the others. 

```sql
CREATE TABLE twitch.throttle_v1 (
    uri text,
    user_id int,
    content text,
    updated_at timestamp,
    PRIMARY KEY ((uri, user_id, content), updated_at)
) WITH CLUSTERING ORDER BY (updated_at ASC)


INSERT INTO throttle_v1 (uri, user_id, content, updated_at) VALUES (?, ?, ?, ?) USING TTL ?
```

Each request can have a dynamic TTL based on the user necessity: 

```rust
impl Throttle {
  pub async fn insert_throttle(&self, connection: &CachingSession, ttl: i32) -> anyhow::Result<()> {
    let query = Query::new(INSERT_THROTTLE_WITH_TTL);

    connection
      .execute(query, (
          &self.uri, 
          &self.user_id,
          &self.content,
          &self.updated_at,
          ttl,
        ),
      )
      .await?;

    Ok(())
  }
}
```

### CODE REVIEW

**Comment by Daniel-Boll** [comment]
@DanielHe4rt it is worth noting that this solution will only work on this endpoint currently, to make it more generic we would have to change the HashMap key to encompass the endpoint too.

File: src/http/v1/metrics_controller.rs [hunk 1]
```diff
@@ -95,6 +98,29 @@ pub async fn post_heartbeat(
...
+    .await?
+    .try_collect()
+    .await?;
+
+  if !throttle_verification.is_empty() {
+    info!("Throttling request: {:?}", throttle);
+    return Ok(HttpResponse::TooManyRequests().finish());
+  }
+
+  info!("Throttle verification: {:?}", throttle_verification);
```
Comment (Daniel-Boll) [inline src/http/v1/metrics_controller.rs:121]: Remove log

File: src/models/v1/throttle.rs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+          &self.uri,
+          &self.user_id,
+          &self.content,
+          &self.updated_at,
+          ttl,
+        ),
+      )
+      .await?;
+
+    Ok(())
```
Comment (Daniel-Boll) [inline src/models/v1/throttle.rs:43]: Shouldn't we use a prepared statement here?

Comment (DanielHe4rt) [reply src/models/v1/throttle.rs:43]: By default all queries inside CachingSession are Prepared. But I'll be adding just in case.

**Pull request merged**